File: IntersectionOfTwoLinkedLists.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param two ListNodes
    # @return the intersected ListNode
    def getIntersectionNode(self, headA, headB):
        m, n = 0, 0
        p, q = headA, headB
        while p:
            p = p.next
            m += 1
        while q:
            q = q.next
            n += 1
        p, q = headA, headB
        if m >= n:
            for i in xrange(m-n):
                p = p.next
        else:
            for i in xrange(n-m):
                q = q.next
                
        while p:
            if p == q:
            #if id(p) == id(q):
                return p
            p = p.next
            q = q.next

        return None

if __name__ == '__main__':
    inter = ListNode(777)
    inter.next = ListNode(999)
    A = ListNode(1)
    A.next = ListNode(3)
    A.next.next = inter
    #B = ListNode(4)
    B = inter

    so = Solution()
    res = so.getIntersectionNode(A,B)
    if res:
        print res.val
    
    

File: CPP/Second/intersection_of_two_linked_lists.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int d = 0;
        for (ListNode *p = headA; p; p = p->next) d++;
        for (ListNode *q = headB; q; q = q->next) d--;
        while (d > 0) {
            headA = headA->next;
            d--;
        }
        while (d < 0) {
            headB = headB->next;
            d++;
        }
        while (headA != headB) {
            headA = headA->next;
            headB = headB->next;
        }
        return headA;
    }
};

int main(int argc, char *argv[])
{

    return 0;
}
